### ISSUE TITLE
per-app and/or global toggle for allowing screenshots (when they disabled by app)

### ISSUE THREAD

**Issue by emanuelb** [GrapheneOS/os-issue-tracker#664]
from twitter thread answers by GrapheneOS: https://twitter.com/GrapheneOS/status/1440759314428297216
> It's important to support it but there's zero problem with adding a way to override the screenshot aspect of it. It just hasn't been implemented. We have no problem with adding a per-app toggle for that or perhaps simply a global toggle though. Someone has to submit clean patch.
> GrapheneOS would have to add a per-app toggle for ignoring it.

looked at how screenshots disable happen by apps:
- using FLAG_SECURE on activity https://developer.android.com/reference/android/view/WindowManager.LayoutParams#FLAG_SECURE
- with setScreenCaptureDisabled Api https://developer.android.com/reference/android/app/admin/DevicePolicyManager.html#setScreenCaptureDisabled%28android.content.ComponentName,%20boolean%29
- is there other ways?

Regarding where and which change to apply in code to override the screenshot disable, not sure, need to search & read more code to know, maybe somewhere in (screenshot/contentcapture related code):
https://github.com/GrapheneOS/platform_frameworks_base/tree/0c0c3352036432e2caea4b1e8957d102906d68d1/core/java/android/view/contentcapture
https://github.com/GrapheneOS/platform_frameworks_base/tree/0c0c3352036432e2caea4b1e8957d102906d68d1/packages/SystemUI/src/com/android/systemui/screenshot

looked at other "solutions" for this problem, like overriding the `setFlags` calls to remove the added SECURE_FLAG:
https://www.securify.nl/en/blog/android-frida-hooking-disabling-flagsecure/ 
https://github.com/veeti/DisableFlagSecure/blob/fd8833a10a1544324f3301f647e74beee9cac58a/src/main/java/fi/veetipaananen/android/disableflagsecure/DisableFlagSecureModule.java
https://forum.xda-developers.com/t/module-smali-patcher-7-4.3680053/

**Comment by System32ita** [comment]
I hope this feature will be implemented soon

**Comment by stevenroose** [comment]
Is there any update available on this? Anyone willing to take it up?

**Comment by stevenroose** [comment]
Does GrapheneOS have an issue bounty system where people can commit a donation for specific issues being implemented?

**Comment by grgpw** [comment]
for some reason its possible to create screenshots in all apps as well as on home- and lockscreen by pressing power and vol down simultaneously.

tested in some Banking apps (DKB, Traderepublic) and ProtonMail, which don't allow regular screenshots.

**Comment by johnnynoone** [comment]
@grgpw Yes, but the protected apps are still blank on the screenshots. Tested with Firefox private mode; in the app itself, in the overview of open apps, and in split screen mode.

**Comment by antonio0831** [comment]
Would be a nice feature. Maybe a per-app toggle will be better?

**Comment by Anti27** [comment]
It seems like the feature is ready, but not released
**https://github.com/GrapheneOS/platform_packages_apps_Settings/pull/209**

**Comment by kernitus** [comment]
> It seems like the feature is ready, but not released **[GrapheneOS/platform_packages_apps_Settings#209](https://github.com/GrapheneOS/platform_packages_apps_Settings/pull/209)**

That PR is huge and has been open for years, not sure how likely it is to get merged?

**Comment by majjuss** [comment]
I would also really like to have this option!

**Comment by LaBlazer** [comment]
Also bumping this

**Comment by Inhishonor** [comment]
Unfortunately, those commits don't apply cleanly anymore. I am currently working on implementing this, by merging it with https://github.com/GrapheneOS/os-issue-tracker/issues/2931 so it can all be controlled by one setting.

**Comment by mio-19** [comment]
For people willing to build the os from source code: To disable flag secure a simple patch might be applied. https://github.com/mio-19/repo/blob/main/Disable-FLAG_SECURE.patch

**Comment by BasedJellyfish11** [comment]
I've opened two PRs which add this functionality with a toggle in developer settings

https://github.com/GrapheneOS/platform_frameworks_base/pull/313
https://github.com/GrapheneOS/platform_packages_apps_Settings/pull/411

@thestinger  Is there a process I should go through to get these reviewed?

**Comment by diamo-dev** [comment]
My preferred implementation (coming from an end user) would be the following;

When taking a screenshot, take one rendering with secure flag, and one without rendering secure flag (what the end user sees), and adding a toggle that switches between the 2, saving the one the user chooses. Separate toggle in screen recording/casting, before the user starts recording/casting.

The toggle for screenshots would be at the following:

<img width="1008" height="2244" alt="Image" src="https://github.com/user-attachments/assets/c38f3598-8572-44b1-a2c5-749754f8a9a0" />

The toggle for screencasting/screen-recording would be here:

<img width="1008" height="2244" alt="Image" src="https://github.com/user-attachments/assets/984d160b-c829-4d46-8017-c30db0c5f914" />

This implementation would most certainly take the most effort, (as compared to adding a developer toggle that simply enables or disables a global flag) and also require some rewrites to the screenshot system.

The idea behind this implementation is there are some apps in which you don't wish to accidentally screenshot/screencast/screen-record the contents of. Such examples could be banking apps or messaging apps.

Making the option a developer toggle makes a lot of people who want to be able to use this feature a lot more vulnerable. My implementation would significantly improve privacy and security, while making it a lot more user friendly.

**Comment by dheera** [comment]
I'd really really like this feature ASAP! This is the one thing keeping me from switching to GrapheneOS. I currently run rooted stock Android + Xposed to disable screenshot blocking as I consider it a privacy and security risk for applications to have control over this. Pixels values sent to a screen I paid money to own are my right to record.

**Comment by thestinger** [comment]
@dheera You don't have any reasonable privacy and security with what you're using as a whole. Please use thumbs up for supporting adding a feature or the issue will have to be locked.

**Comment by aurocha** [comment]
Just want to bump so it isn't forgotten. Thanks ❤️

**Comment by ijustworkhereshrug** [comment]
This would be an excellent feature

**Comment by thestinger** [comment]
Please don't post comments in support of a feature. Use the thumbs up reaction.

**Comment by jason-eu** [comment]
> Please don't post comments in support of a feature. Use the thumbs up reaction.

It's a 108 thumbs up with no real word on when there will be an implementation, just let them comment mate

**Comment by schklom** [comment]
@jason-eu That will just force them to lock this issue, just to stop getting pointless emails.

**Comment by TopchetoEU** [comment]
this has been an issue for over 4 years and yet no real progress has been made on it. what the actual hell. im well-aware of the hardships of a foss project, but by now, at the very least we should've gotten clear communication about where this issue is going

**Comment by thestinger** [comment]
@TopchetoEU We've made tons of progress on many things. There are too many planned features for everything to get quickly implemented. This isn't very important and hasn't been a high priority. Would you rather have this of Location Scopes, Camera Scopes, Microphone Scopes, VPN leak fixes and other privacy fixes? If you really need to then you can take a picture of a phone with another.

**Comment by ijustworkhereshrug** [comment]
Yes.  I think Graphene does a lot of things very well.  The Scoping is great, but this is a capability that is giving control back to the user.
